### PR TITLE
tests(#473): add missing tests for send with, no send flow.

### LINKED ISSUE #473: [Task] Support whole noSend-noSend-...-sendWith scenario

### PR DESCRIPTION
**Pull request by mgosek-4chain**
## Linked Issues / Tickets

image

Reference any related issues or tickets, e.g. "Closes #473".

## Testing Procedure

Describe the tests you've added or any testing steps you've taken.

- [X] I have added new unit tests
- [X] All tests pass locally
- [X] I have tested manually in my local environment

## Checklist:

- [X] I have performed a self-review of my own code
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have run the linter

### CODE REVIEW

File: pkg/internal/validate/validate_process_action_args.go [hunk 1]
```diff
@@ -23,10 +23,6 @@ func ProcessActionArgs(args *wdk.ProcessActionArgs) error {
 		}
 	}
 
-	if !args.IsNoSend && args.TxID == nil {
```
Comment (chris-4chain) [inline pkg/internal/validate/validate_process_action_args.go:26]: Perhaps we should keep this check but add also `!isSendWith` to the condition.

File: pkg/storage/provider_no_send_test.go [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+	keyDeriver := testusers.Alice.KeyDeriver(t)
+	activeStorage := given.Provider().GORM()
+
+	given.
+		Action(activeStorage).
+		WithSatoshisToInternalize(inputSatoshish).
+		WithSatoshisToSend(1).
+		Processed()
+
+	// Step 1 - Process Action with IsNoSend flag set to true, empty noSendChange outpoints. Inputs will be allocated
```
Comment (chris-4chain) [inline pkg/storage/provider_no_send_test.go]: In another tests where we test a long story, we have separated `t.Run` to define steps. Consider it here. 

Or, you can provide some testability to make the story shorter.

File: pkg/storage/provider_no_send_test.go [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+	}
+
+	thirdCreateActionResult, err := activeStorage.CreateAction(t.Context(), testusers.Alice.AuthID(), createActionArgs)
+	require.NoError(t, err)
+	require.NotNil(t, thirdCreateActionResult)
+
+	thirdTx, err := assembler.NewCreateActionTransactionAssembler(keyDeriver, nil, secondCreateActionResult).Assemble()
+	require.NoError(t, err)
+	require.NotNil(t, secondTx)
+	require.NoError(t, secondTx.Sign())
```
Comment (chris-4chain) [inline pkg/storage/provider_no_send_test.go]: There is an error here. You need to check and sign `thirdTx`.  
This is crucial. ⚠️

File: pkg/storage/provider_no_send_test.go [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+	"github.com/stretchr/testify/require"
+)
+
+func TestProcessAction_SendFlagFalse_SendWithSlice_IncludesTwoTransactions(t *testing.T) {
+	// given:
+	given, cleanup := testabilities.Given(t)
+	defer cleanup()
+
+	// and:
+	const inputSatoshish = 99904
```
Comment (chris-4chain) [inline pkg/storage/provider_no_send_test.go]:
```suggestion
	const inputSatoshis = 99904
```

**Review by ac4ch — APPROVED** [review]

File: pkg/internal/fixtures/default_valid_create_action_args.go [hunk 5]
```diff
@@ -13,8 +13,40 @@ import (
 
 const DefaultCreateActionOutputSatoshis = 42000
 
-func DefaultValidCreateActionArgs() wdk.ValidCreateActionArgs {
-	return wdk.ValidCreateActionArgs{
+func DefaultWalletCreateActionArgs(t *testing.T, opts ...func(*sdk.CreateActionArgs)) sdk.CreateActionArgs {
```
Comment (dorzepowski) [inline pkg/internal/fixtures/default_valid_create_action_args.go:16]: ### minor thing 
Why you changed the order of functions here? It's harder to track the changes you made

File: pkg/wallet/wallet_create_action_test.go [hunk 6]
```diff
@@ -693,36 +691,81 @@ func (s *WalletTestSuite) TestWalletCreateAction_NoSend_SendWith() {
...
 
 		// when:
-		args = fixtures.DefaultWalletCreateActionArgs(t, walletargs.WithSendWith(firstResult.Txid))
+		args = fixtures.DefaultWalletCreateActionArgs(t,
+			walletargs.WithNoSendChangeOutputs(firstResult.NoSendChange...),
+			walletargs.WithSendWith(firstResult.Txid),
+			walletargs.WithSatoshisAsFirstOutput(1),
+		)
+
+		_, err = aliceWallet.CreateAction(t.Context(), args, fixtures.DefaultOriginator)
```
Comment (dorzepowski) [inline pkg/wallet/wallet_create_action_test.go]: for the sake of keeping our convention (that we return empty result when there is an error) I would add `assert.Nil(result)`

File: pkg/wallet/wallet_create_action_test.go [hunk 7]
```diff
@@ -693,36 +691,81 @@ func (s *WalletTestSuite) TestWalletCreateAction_NoSend_SendWith() {
...
+		thirdResult, err := aliceWallet.CreateAction(t.Context(), args, fixtures.DefaultOriginator)
+
+		// then:
+		require.NoError(t, err)
 
-		assert.Equal(t, secondResult.SendWithResults[0].Txid, firstResult.Txid, "Wallet result should have same txid as the one from first send with result")
-		assert.Equal(t, secondResult.SendWithResults[0].Status, sdk.ActionResultStatusUnproven, "Wallet send with result should have unproven status")
-		assert.Equal(t, secondResult.SendWithResults[1].Txid, secondResult.Txid, "Wallet result should have same txid as the one from second send with result")
-		assert.Equal(t, secondResult.SendWithResults[1].Status, sdk.ActionResultStatusUnproven, "Wallet send with result should have unproven status")
+		//// and:
```
Comment (dorzepowski) [inline pkg/wallet/wallet_create_action_test.go]: ## very minor
```suggestion
		// and:
```

File: pkg/wallet/wallet_create_action_test.go [hunk 8]
```diff
@@ -693,36 +691,81 @@ func (s *WalletTestSuite) TestWalletCreateAction_NoSend_SendWith() {
...
+		assert.Equal(t, thirdResult.SendWithResults[0].Txid, firstResult.Txid, "Wallet result should have same txid as the one from first send with result")
+		assert.Equal(t, thirdResult.SendWithResults[0].Status, sdk.ActionResultStatusUnproven, "Wallet send with result should have unproven status")
+		assert.Equal(t, thirdResult.SendWithResults[1].Txid, secondResult.Txid, "Wallet result should have same txid as the one from second send with result")
+		assert.Equal(t, thirdResult.SendWithResults[1].Status, sdk.ActionResultStatusUnproven, "Wallet send with result should have unproven status")
```
Comment (dorzepowski) [inline pkg/wallet/wallet_create_action_test.go:769]: maybe we should check also wallet state, like it is done for example in TestWalletCreateAction_SignableTx

File: pkg/storage/provider_no_send_test.go [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+	"github.com/bsv-blockchain/go-wallet-toolbox/pkg/internal/fixtures/testusers"
+	"github.com/bsv-blockchain/go-wallet-toolbox/pkg/storage/internal/testabilities"
+	"github.com/bsv-blockchain/go-wallet-toolbox/pkg/wdk"
+	"github.com/bsv-blockchain/go-wallet-toolbox/pkg/wdk/primitives"
+	"github.com/go-softwarelab/common/pkg/slices"
+	"github.com/go-softwarelab/common/pkg/to"
+	"github.com/stretchr/testify/require"
+)
+
+func TestNoSendSendWithScenario_SendWithoutNewTx_2(t *testing.T) {
```
Comment (dorzepowski) [inline pkg/storage/provider_no_send_test.go]: please remove `_2` suffix from all tests here

File: pkg/internal/validate/valid_create_action_args.go [hunk 10]
```diff
@@ -91,6 +83,19 @@ func WalletCreateActionArgs(args *wdk.ValidCreateActionArgs) error {
 			return fmt.Errorf("invalid output at index %d: %w", i, err)
 		}
 	}
+
+	if !args.IsNewTx && len(args.Options.SendWith) == 0 {
```
Comment (dorzepowski) [inline pkg/internal/validate/valid_create_action_args.go:87]: do we have tests for all those new validations?

File: pkg/internal/fixtures/default_valid_create_action_args.go (hunk 5 above)
Comment (mgosek-4chain) [reply pkg/internal/fixtures/default_valid_create_action_args.go:16]: It wasn't my intention to do that, so I'd like to apologize :). I was working on a concept, and some parts were commented out. After commenting and removing my ideas, I eventually lost the original ordering ^^

File: pkg/wallet/wallet_create_action_test.go (hunk 6 above)
Comment (mgosek-4chain) [reply pkg/wallet/wallet_create_action_test.go]: Done.

File: pkg/wallet/wallet_create_action_test.go (hunk 7 above)
Comment (mgosek-4chain) [reply pkg/wallet/wallet_create_action_test.go]: Done.

File: pkg/wallet/wallet_create_action_test.go (hunk 8 above)
Comment (mgosek-4chain) [reply pkg/wallet/wallet_create_action_test.go:769]: Done.

File: pkg/storage/provider_no_send_test.go (hunk 9 above)
Comment (mgosek-4chain) [reply pkg/storage/provider_no_send_test.go]: Done.

File: pkg/internal/validate/valid_create_action_args.go (hunk 10 above)
Comment (mgosek-4chain) [reply pkg/internal/validate/valid_create_action_args.go:87]: I have added the missing edge case.

Comment (dorzepowski) [reply pkg/internal/validate/valid_create_action_args.go:87]: Maybe I'm wrong but I don't see a test for duplicates in NoSendChange

**Review by dorzepowski — APPROVED** [review]

**Review by chris-4chain — APPROVED** [review]
